q2_metadata/normalization/_norm_check_rules.py

```diff
@@ -0,0 +1,311 @@
+            return True
+    return False
+
+
+def check_validation(rule_value):
+    """Check that the user-defined value for the
+    rule "validation" is correctly formatted.
+
+    validation must be a dict
+        Apply the rule of validating variables values: replace by the blank
+        (or the missing) value every sample entry that for the current variable has
+        in one or multiple other variable(s) a null or missing value for that sample.
+                    validation:
+                      force_to_blank_if:
+                        is null:
+                        - variable
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "validation".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    # accepted structure for the "validation" rule
+    possible_validations = {
+        'force_to_blank_if': {'is null'}
+    }
+    # check the first level is a dict
+    if isinstance(rule_value, dict):
+        # get the current keys
+        rule_value_set = set(rule_value)
+        # get the accepted keys from the current keys
+        accepted_set = rule_value_set & set(possible_validations)
+        # if there is not more keys than the accepted ones
+        if not len(rule_value_set ^ accepted_set):
+            # get the possible conditions
+            possible_validations_2 = possible_validations['force_to_blank_if']
+            # for each of the condition/replacements for the accepted keys
+            for rule_2, rule_value_2 in rule_value['force_to_blank_if'].items():
+                # check that the content of replacements are lists
+                if isinstance(rule_value_2, list):
+                    # get the accepted conditions from the current conditions
+                    if rule_2 in possible_validations_2:
+                        return True
+    return False
+
+
+def check_normalization(rule_value):
+    """Check that the user-defined value for the
+    rule "normalization" is correctly formatted.
+
+    normalization must be a dict
+        Apply the rules of normalizing the variables values, by replacing
+        the values outside the range by a given replacement value.
+                    normalization:
+                        gated_value:
+                            - Out of bounds
+                        maximum:
+                            - int
+                            - float
+                        minimum:
+                            - int
+                            - float
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "normalization".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    if isinstance(rule_value, dict):
+        possible_normalizations = {'maximum', 'minimum', 'gated_value'}
+        rule_value_set = set(rule_value)
+        if not len(rule_value_set ^ (rule_value_set & possible_normalizations)):
```

Comment:
Here you want to check that all three values are present, right? Wouldn't it be easier to check if the sets are equal?

Reply:
I believe this is "if there exist `possible_normalizations` that ARE NOT in the `rule_value_set`, return `True`". This is based on:

```python
>>> {1, 2, 3} ^ ({1, 2, 3} & {3, 4, 5})
{1, 2}
```

...and `not len(<set>)` evaluating `True` when the cardinality is zero. 

I don't understand how this has any relation to the docstring.

Reply:
This  "if there exist rule_value_set that ARE NOT in the possible_normalizations, return True".
but this is also tackled with the much simpler expression proposed above: `rule_value_set.issubset(possible_normalizations)` which required that all the items in rule_value_set are part of possible_normalizations, and not other extraneous items is allowed.

The doctring was a remnant of a very early version:
```
ef check_normalization(rule_value):
    """Check that the user-defined value for the rule
    "normalization" is a dictionary which keys are no
    other than "minimum", "maximum" or "gated_value",
    and which values are numeric.

    Parameters
    ----------
    rule_value
        Parsed rule value for the rule "normalization".

    Returns
    -------
    bool
        True if successful, False otherwise.
    """
    allowed = {'maximum', 'minimum', 'gated_value'}
    rule_value_set = set(rule_value)
    if isinstance(rule_value, dict) and rule_value_set.issubset(allowed):
        is_int_or_float = map(lambda x: isinstance(x, (int, float)), rule_value.values())
        if len(rule_value) == sum(is_int_or_float):
            return True
    return False

```